matrix.go

```diff
@@ -8,42 +8,64 @@ import (
 	"net/url"
 )
 
+type Params map[string][]string
+type ParamKeys []string
+
 type Matrix struct {
 	Clients   []string
-	Servers   []string
-	Behaviors []string
+	ParamKeys ParamKeys
+	Params    Params
 }
 
 type Results []Result
 
 func BeginMatrixTest(matrix Matrix) Results {
 
-	len := len(matrix.Clients) * len(matrix.Servers) * len(matrix.Behaviors)
-	results := make(Results, 0, len)
+	testCases := []TestCase{}
 
 	for _, client := range matrix.Clients {
-		for _, server := range matrix.Servers {
-			for _, behavior := range matrix.Behaviors {
-
-				testCase := TestCase{
-					Client:   client,
-					Server:   server,
-					Behavior: behavior,
-				}
-
-				result := ExecuteTestCase(testCase)
-				results = append(results, result)
-			}
-		}
+		collect(&testCases, client, matrix, 0, make(map[string]string, 0))
+	}
+
+	// TODO - the testCases end up being the same here, something about the pointer usecase is fucked
+	// suggest just switching to channels at this point...
+
+	results := make(Results, 0)
+	for _, testCase := range testCases {
+
+		fmt.Println("TESTCASE", testCase)
+
+		result := ExecuteTestCase(testCase)
+		results = append(results, result)
 	}
 
 	return results
 }
 
+func collect(testCases *[]TestCase, client string, matrix Matrix, index int, params map[string]string) {
+
+	if index == len(matrix.ParamKeys) {
+		testCase := TestCase{
+			Client: client,
+			Params: params,
+		}
+
+		*testCases = append(*testCases, testCase)
```

Comment:
@breerly  The issue is that you're using the same `params` in all the test cases instead of a copy. Since they all point to the same map, the last values you set in params is what every test case sees.

Also, pointer to slice is very rarely used, you might want to just return slices from the function, append them all to a single larger slice and return that.